contracts/Liquidator.sol

```diff
@@ -79,35 +82,48 @@ contract Liquidator is WithAdmin, ReentrancyGuard {
     /// @notice Emitted when a liquidator is removed from the allowlist
     event AllowlistEntryRemoved(address indexed borrower, address indexed liquidator);
 
-    using SafeERC20 for IERC20;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
 
-    constructor(
-        address admin_,
-        address payable vBnb_,
-        address comptroller_,
-        address vaiController_,
-        address treasury_,
-        uint256 treasuryPercentMantissa_
-    )
-        WithAdmin(admin_)
-        ReentrancyGuard()
-    {
-        ensureNonzeroAddress(admin_);
+    /// @notice Constructor for the implementation contract. Sets immutable variables.
+    /// @param comptroller_ The address of the Comptroller contract
+    /// @param vBnb_ The address of the VBNB
+    /// @param treasury_ The address of Venus treasury
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address comptroller_, address payable vBnb_, address treasury_) {
         ensureNonzeroAddress(vBnb_);
         ensureNonzeroAddress(comptroller_);
-        ensureNonzeroAddress(vaiController_);
         ensureNonzeroAddress(treasury_);
         vBnb = IVBNB(vBnb_);
         comptroller = IComptroller(comptroller_);
-        vaiController = IVAIController(vaiController_);
+        vaiController = IVAIController(IComptroller(comptroller_).vaiController());
         treasury = treasury_;
+        _disableInitializers();
+    }
+
+    /// @notice Initializer for the implementation contract.
+    /// @param treasuryPercentMantissa_ Treasury share, scaled by 1e18 (e.g. 0.2 * 1e18 for 20%)
+    function initialize(uint256 treasuryPercentMantissa_) external virtual initializer {
+        __Liquidator_init(treasuryPercentMantissa_);
+    }
+
+    /// @dev Liquidator initializer for derived contracts.
+    /// @param treasuryPercentMantissa_ Treasury share, scaled by 1e18 (e.g. 0.2 * 1e18 for 20%)
+    function __Liquidator_init(uint256 treasuryPercentMantissa_) internal onlyInitializing {
+        __Ownable2Step_init();
+        __ReentrancyGuard_init();
+        __Liquidator_init_unchained(treasuryPercentMantissa_);
+    }
+
+    /// @dev Liquidator initializer for derived contracts that doesn't call parent initializers.
+    /// @param treasuryPercentMantissa_ Treasury share, scaled by 1e18 (e.g. 0.2 * 1e18 for 20%)
+    function __Liquidator_init_unchained(uint256 treasuryPercentMantissa_) internal onlyInitializing {
```

Comment:
why we have separated this logic from `__Liquidator_init ` function?

Reply:
Just following OpenZeppelin convention for initializers. According to this convention, upgradeable contracts have two internal initializers – a regular one that calls the base initializers, and the "unchained" one that doesn't call base initializers. This is done for the "diamond inheritance" cases, e.g. if you have two base contracts, both inheriting "OwnableUpgradeable".

It's not really applicable here since no contract will ever inherit both Liquidator and Ownable2StepUpgradeable simultaneously, but they use this convention in all their contracts, so I decided to do the same 🤷

Reply:
yeah better follow conventions thanks